### DISCUSSION TITLE
federated SPARQL query planning

### DISCUSSION THREAD

**Discussion by balhoff** [comunica/comunica#1217] [Q&A] [answered]
I've been exploring federated SPARQL query in Comunica—really impressive so far. I have a question about how it decides which queries to submit to SPARQL endpoints. It seems that when you have multiple SPARQL endpoint datasources, it treats them like triple pattern fragment servers, submitting single-triple queries (which makes sense). But it seems to make some unnecessary requests. The example I'm talking about is [here](https://query.linkeddatafragments.org/#datasources=https%3A%2F%2Fquery.wikidata.org%2Fsparql;https%3A%2F%2Fsparql.uniprot.org%2Fsparql&query=PREFIX%20rdfs%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0APREFIX%20wd%3A%20%3Chttp%3A%2F%2Fwww.wikidata.org%2Fentity%2F%3E%0APREFIX%20wdt%3A%20%3Chttp%3A%2F%2Fwww.wikidata.org%2Fprop%2Fdirect%2F%3E%0ASELECT%20*%0AWHERE%20%7B%0A%20%20wd%3AQ50265665%20wdt%3AP683%20%3Fchebi_id%20.%0A%20%20BIND(IRI(CONCAT(%22http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FCHEBI_%22%2C%20%3Fchebi_id))%20AS%20%3Fchebi)%0A%20%20%3Fchebi%20rdfs%3AsubClassOf%20%3Fsuperclass%20.%0A%7D%0ALIMIT%2010).

I added two SPARQL endpoints:
- `https://query.wikidata.org/sparql`
- `https://sparql.uniprot.org/sparql`

This is my query:

```sparql
PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
PREFIX wd: <http://www.wikidata.org/entity/>
PREFIX wdt: <http://www.wikidata.org/prop/direct/>
SELECT *
WHERE {
  wd:Q50265665 wdt:P683 ?chebi_id .
  BIND(IRI(CONCAT("http://purl.obolibrary.org/obo/CHEBI_", ?chebi_id)) AS ?chebi)
  ?chebi rdfs:subClassOf ?superclass .
}
LIMIT 10
```

The first triple pattern exists in Wikidata, and the second exists in UniProt. These are the queries submitted by the SPARQL engine:
1. UniProt: `SELECT (COUNT(*) AS ?count) WHERE { ?chebi <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **70666020**
2. UniProt: `SELECT (COUNT(*) AS ?count) WHERE { <http://www.wikidata.org/entity/Q50265665> <http://www.wikidata.org/prop/direct/P683> ?chebi_id. }` -> **0**
3. UniProt: `SELECT ?chebi ?superclass WHERE { ?chebi <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **More than 5 GB data**
4. UniProt: `SELECT ?chebi_id WHERE { <http://www.wikidata.org/entity/Q50265665> <http://www.wikidata.org/prop/direct/P683> ?chebi_id. }` -> **empty**
5. Wikidata: `SELECT (COUNT(*) AS ?count) WHERE { ?chebi <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **0**
6. Wikidata: `SELECT (COUNT(*) AS ?count) WHERE { <http://www.wikidata.org/entity/Q50265665> <http://www.wikidata.org/prop/direct/P683> ?chebi_id. }` -> **1**
7. Wikidata: `SELECT ?chebi ?superclass WHERE { ?chebi <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **empty**
8. Wikidata: `SELECT ?chebi_id WHERE { <http://www.wikidata.org/entity/Q50265665> <http://www.wikidata.org/prop/direct/P683> ?chebi_id. }` -> **one row**
9. Wikidata: `SELECT (COUNT(*) AS ?count) WHERE { <http://purl.obolibrary.org/obo/CHEBI_5931> <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **0**
10. UniProt: `SELECT (COUNT(*) AS ?count) WHERE { <http://purl.obolibrary.org/obo/CHEBI_5931> <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **4**
11. Wikidata: `SELECT ?superclass WHERE { <http://purl.obolibrary.org/obo/CHEBI_5931> <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **empty**
12. UniProt: ` SELECT ?superclass WHERE { <http://purl.obolibrary.org/obo/CHEBI_5931> <http://www.w3.org/2000/01/rdf-schema#subClassOf> ?superclass. }` -> **4 rows**

Why does it not avoid executing query 3, which unnecessarily downloads 5 GB of data, especially since it later executes query 12? And why submit any of the SELECT queries when the corresponding COUNT query is `0`?

I tested the same query on two triple pattern fragments servers running locally, backed by these two SPARQL endpoints. The result is more like what I expect; it avoids downloading a huge amount of data from UniProt, and runs more quickly. Thanks—Comunica looks like a great platform.

**Comment by rubensworks** [comment, marked as answer]
Hi @balhoff, thanks for the detailed findings!

You're right to indicate that if a count query returns 0, that no following SELECT query should be initiated.

Internally, Comunica will fetch the COUNT and SELECT queries at the same time.
However, the SELECT query should be lazy, and only be actually triggered from the moment it actually becomes required.
As you have indicated, something seems to be going wrong with this lazyness, which is making the query always get triggered.

The precise location where this lazy stream is created is here: https://github.com/comunica/comunica/blob/master/packages/actor-rdf-resolve-hypermedia-sparql/lib/RdfSourceSparql.ts#L184
So either this stream is not actually lazy, or somewhere else in the codebase (e.g. the join operators), the stream is being triggered too early.

I'll convert this discussion into an issue, as it's a bug.

**Reply by balhoff** [reply]
Great, thank you @rubensworks.
